File: fixtures/java/j01/tests/src/test/java/UsersTest.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class UsersTest {

    @BeforeEach
    void setUp() {
        // Reset users list to initial state before each test
        Fixture.users.clear();
        Fixture.users.add(new HashMap<>(Map.of("id", 1, "name", "Alice Smith", "email", "alice@example.com")));
        Fixture.users.add(new HashMap<>(Map.of("id", 2, "name", "Bob Jones", "email", "bob@example.com")));
    }

    @Test
    void createsUserWithValidDataAndReturns201() {
        Users.Result res = Users.createUser("Charlie Brown", "charlie@example.com", "Password1");
        assertEquals(201, res.status);
        assertNotNull(res.body);
        assertNotNull(res.body.get("id"));
        assertEquals("Charlie Brown", res.body.get("name"));
        assertEquals("charlie@example.com", res.body.get("email"));
    }

    @Test
    void doesNotReturnPasswordInResponse() {
        Users.Result res = Users.createUser("Dave Example", "dave@example.com", "Secure123");
        assertEquals(201, res.status);
        assertFalse(res.body.containsKey("password"));
    }

    @Test
    void returns400WhenNameIsMissing() {
        Users.Result res = Users.createUser(null, "missing@example.com", "Password1");
        assertEquals(400, res.status);
        assertNotNull(res.errors);
        assertTrue(res.errors.size() > 0);
    }

    @Test
    void returns400WhenNameIsTooShort() {
        Users.Result res = Users.createUser("A", "short@example.com", "Password1");
        assertEquals(400, res.status);
        assertNotNull(res.errors);
        assertTrue(res.errors.size() > 0);
    }

    @Test
    void returns400WhenEmailIsInvalidFormat() {
        Users.Result res = Users.createUser("Valid Name", "not-an-email", "Password1");
        assertEquals(400, res.status);
        assertNotNull(res.errors);
        assertTrue(res.errors.size() > 0);
    }

    @Test
    void returns400WhenPasswordIsTooShort() {
        Users.Result res = Users.createUser("Valid Name", "valid@example.com", "abc");
        assertEquals(400, res.status);
        assertNotNull(res.errors);
        assertTrue(res.errors.size() > 0);
    }

    @Test
    void returns400WhenPasswordHasNoUppercaseLetter() {
        Users.Result res = Users.createUser("Valid Name", "valid2@example.com", "password1");
        assertEquals(400, res.status);
        assertNotNull(res.errors);
        assertTrue(res.errors.size() > 0);
    }

    @Test
    void returns409WhenEmailAlreadyExists() {
        Users.Result res = Users.createUser("Alice Copy", "alice@example.com", "Password1");
        assertEquals(409, res.status);
        assertNotNull(res.message);
        assertTrue(res.message.toLowerCase().contains("already exists"));
    }
}

File: fixtures/nodejs/j01/tests/users.test.js
const request = require('supertest');
const { app } = require('./users');

describe('POST /users', () => {
  test('creates user with valid data and returns 201', async () => {
    const res = await request(app).post('/users').send({
      name: 'Charlie Brown',
      email: 'charlie@example.com',
      password: 'Password1',
    });
    expect(res.status).toBe(201);
    expect(res.body).toHaveProperty('id');
    expect(res.body.name).toBe('Charlie Brown');
    expect(res.body.email).toBe('charlie@example.com');
  });

  test('does not return password in response', async () => {
    const res = await request(app).post('/users').send({
      name: 'Dave Example',
      email: 'dave@example.com',
      password: 'Secure123',
    });
    expect(res.status).toBe(201);
    expect(res.body).not.toHaveProperty('password');
  });

  test('returns 400 when name is missing', async () => {
    const res = await request(app).post('/users').send({
      email: 'missing@example.com',
      password: 'Password1',
    });
    expect(res.status).toBe(400);
    expect(Array.isArray(res.body)).toBe(true);
    expect(res.body.length).toBeGreaterThan(0);
  });

  test('returns 400 when name is too short (1 char)', async () => {
    const res = await request(app).post('/users').send({
      name: 'A',
      email: 'short@example.com',
      password: 'Password1',
    });
    expect(res.status).toBe(400);
    expect(Array.isArray(res.body)).toBe(true);
  });

  test('returns 400 when email is invalid format', async () => {
    const res = await request(app).post('/users').send({
      name: 'Valid Name',
      email: 'not-an-email',
      password: 'Password1',
    });
    expect(res.status).toBe(400);
    expect(Array.isArray(res.body)).toBe(true);
  });

  test('returns 400 when password is too short', async () => {
    const res = await request(app).post('/users').send({
      name: 'Valid Name',
      email: 'valid@example.com',
      password: 'abc',
    });
    expect(res.status).toBe(400);
    expect(Array.isArray(res.body)).toBe(true);
  });

  test('returns 400 when password has no uppercase letter', async () => {
    const res = await request(app).post('/users').send({
      name: 'Valid Name',
      email: 'valid2@example.com',
      password: 'password1',
    });
    expect(res.status).toBe(400);
    expect(Array.isArray(res.body)).toBe(true);
  });

  test('returns 409 when email already exists', async () => {
    const res = await request(app).post('/users').send({
      name: 'Alice Copy',
      email: 'alice@example.com',
      password: 'Password1',
    });
    expect(res.status).toBe(409);
    expect(res.body).toHaveProperty('message');
    expect(res.body.message).toMatch(/already exists/i);
  });
});
